### DISCUSSION TITLE
Memory usage of the WebUI

### DISCUSSION THREAD

**Discussion by 9SMTM6** [open-webui/open-webui#2583] [General]
# Bug Report

## Description

**Bug Summary:**
Open WebUI uses a lot of RAM, IMO without reason.

I'm currently running the WebUI on a Raspberry, to have my chats always available and for security - i can keep traffic with my reverse proxy on device -, ollama runs on another PC.
However I'm thinking if I can keep it that way, since the container, even from a fresh start, uses a HUGE amount of memory. If I check memory usage of containers with `docker container stats` then Open WebUI uses 500MB from a cold start, and often it seems to use more than 1 GB, and all that is with LiteLLM disabled.

My next largest memory using container is Gitea, which sips just 100MB.

WebUI also seems to have 15 worker processes or similar running all the time.

I feel like thats totally overkill and not required. And I can not find any environment variable or similar where I could configure WebUI to be less aggressive when it comes to resource usage.

**Steps to Reproduce:**
Start the container (ghcr.io/open-webui/open-webui:main) on docker (in my case with ENABLE_LITELLM=false, but I cant imagine it being better without), observe usage with `docker container stats`, compare with other services, eg. Gitea. 

**Expected Behavior:**
I can see the container maybe having high usage if you actively do RAG or similar, but not while idling

**Actual Behavior:**
spawns 15 processes (all look identical in my system level htop), eats half a gigabyte of RAM without me having opened the site once, I've observed the container eating more than a Gigabyte.

## Environment

- **Open WebUI Version:** v0.1.125
- **Ollama (if applicable):** 0.1.38

- **Operating System:** Arch ARM on Raspberry Pi 4
- **Browser (if applicable):** not applicable

## Reproduction Details

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [x] I have included the Docker container logs.

## Logs and Screenshots

**Browser Console Logs:**
not applicable

**Docker Container Logs:**
Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.
Generating WEBUI_SECRET_KEY
Loading WEBUI_SECRET_KEY from .webui_secret_key
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)

  ___                    __        __   _     _   _ ___ 
 / _ \ _ __   ___ _ __   \ \      / /__| |__ | | | |_ _|
| | | | '_ \ / _ \ '_ \   \ \ /\ / / _ \ '_ \| | | || | 
| |_| | |_) |  __/ | | |   \ V  V /  __/ |_) | |_| || | 
 \___/| .__/ \___|_| |_|    \_/\_/ \___|_.__/ \___/|___|
      |_|                                               

      
v0.1.125 - building the best open-source AI user interface.      
https://github.com/open-webui/open-webui

INFO:     127.0.0.1:56220 - "GET /health HTTP/1.1" 200 OK

**Screenshots (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

## Installation Method

Docker, behind traefik.

## Additional Information

thats all.

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by 9SMTM6** [comment]
I tried setting WEB_CONCURRENCY [from uvicorn](https://www.uvicorn.org/settings/#production), but that made it worse (you need to delete the container and redeploy for that variable to make any difference at all).

I'm not sure why uvicorn starts so many processes /or rather why they use so much RAM, gitea also has a bunch of processes that show up in htop.

**Comment by 9SMTM6** [comment]
UVICORN_LIMIT_CONCURRENCY (https://www.uvicorn.org/settings/ + https://www.uvicorn.org/settings/#resource-limits) has no effect.

**Comment by LumenYoung** [comment]
Yes, I think this is also wierd, when comparable services uaually takes only ~ 20 MiB of ram while Open-WebUI takes 500 MiB!

Don't know if we can trace down to the cultprit for this consumption

**Comment by FotisK** [comment]
on my macOS machine it exhausts the 8GB that I have allocated to docker as soon as I restart the container

**Reply by justinh-rahb** [reply]
WebUI isn't fat, it's just stress eating!

/s

Ya... it's ballooned a bit. We pack a lot of functionality in...

**Reply by 9SMTM6** [reply]
But honestly immediately eating up 8GB after a restart is way worse than what I saw.

Unless you use a container with bundled ollama and something is making it load the LLM immediately? IDK how unified RAMs 'VRAM' usage shows up on Macs task manager.

If that's not it. IDK. Something else is going on. Is webui loading some saved data into memory, like eg. past chat logs etc?

@justinh-rahb making that added functionality configurable would probably be a start.

**Reply by justinh-rahb** [reply]
Upon restart there shouldn't really be much loaded, we lazy-load the models now. If you use RAG or Whisper those models will get loaded, but they're not huge. Not unless you change them anyway.

Looking at my WebUI container just now, idling at 1.4GB.. which is definitely not small, but also not huge 🤷‍♂️

![Screenshot 2024-07-26 at 6 29 09 PM](https://github.com/user-attachments/assets/0ea99f6b-b089-42ad-995b-3642b899297d)

**Reply by LumenYoung** [reply]
> Upon restart there shouldn't really be much loaded, we lazy-load the models now. If you use RAG or Whisper those models will get loaded, but they're not huge. Not unless you change them anyway.
> 
> Looking at my WebUI container just now, idling at 1.4GB.. which is definitely not small, but also not huge 🤷‍♂️
> 
> ![Screenshot 2024-07-26 at 6 29 09 PM](https://github.com/user-attachments/assets/0ea99f6b-b089-42ad-995b-3642b899297d)
> 

Well, i would say it is realllly huge when you keep in mind that most of the servers people buy to self host usually have 1-2gbs of ram. And as per my experience of deploying so many services on my server, usually the web service like webui would cost only less than 50 MB.🙈

**Reply by justinh-rahb** [reply]
> Well, i would say it is realllly huge when you keep in mind that most of the servers people buy to self host usually have 1-2gbs of ram.

Speak for yourself lol

![Screenshot 2024-07-26 at 7 00 00 PM](https://github.com/user-attachments/assets/b900f61f-a4f1-4e24-a35c-4cf0abf86005)

Alright I get it, RAM doesn't grow on trees, but 1-2GB? What year is it? Bill G is that you..? Come on 512k—or 1GB, is _not_ enough for anyone anymore and hasn't been for a long time. That's what I allocate for my _smallest_ production web app containers or cloud VMs. WebUI is simply, not small. We've contemplated at various times and still haven't ruled out the possibility of providing options to break the container up into more modular pieces that can be optional, but that's development work that isn't _adding features_ per se, so it's not as high priority right now.

**Reply by FotisK** [reply]
in my particular case, since I'm on an M1 mac with 16GB of shared RAM/GPU memory, that 1 or 2GB overhead make the difference between being able to run an 8b-12b param model and not being able to run it at all.
I run docker only for open webui (ollama runs natively on my mac) and now that I limited the container to 2GB of RAM it's not really complaining and is idling at 1,5GB out of the allocated 2GB. It doesn't help that docker desktop for mac has its own memory overhead though.

**Reply by justinh-rahb** [reply]
@FotisK btw.. was on my phone when I read this originally and now seeing it again on a desktop I noticed more context.. if you're seeing 8GB in Activity Monitor on macOS for a Docker application... that might be because it's running in a Linux VM that gets around 8GB of RAM allocated to it.

**Reply by FotisK** [reply]
> @FotisK btw.. was on my phone when I read this originally and now seeing it again on a desktop I noticed more context.. if you're seeing 8GB in Activity Monitor on macOS for a Docker application... that might be because it's running in a Linux VM that gets around 8GB of RAM allocated to it.

right! Part of the overhead is definitely the linux vm!

**Comment by 9SMTM6** [comment]
@justinh-rahb 1.4GB at idle is pretty damn huge for something that's just sitting there.

I would be able to justify something like that if I were to serve a bunch of people, but as a matter of fact it's mostly just me.

Honestly I'm surprised how unfazed by this evidently most people are. I guess most people host this on the server they host the LLM, and if you're doing that then this container doesn't make much of a difference anyways.

But trust me. It's huge for a webui.

I don't really get why it's doing that. If you always load some models like the RAG you mentioned then perhaps that's some explanation. Because that memory amount doesn't come from like it simply being a python app.

I run this on my raspberry. The LLMs are on my powerful PC, that has wake on lan configured, so I can start and use it from anywhere (ssh jump host), but it's off most of the time. Having the WebUI on that host makes a few things easier, but most importantly, it means that I've always got the chats available to reference. 

I've got a a shitton of services running on that raspberry. Nothing else has comparable RAM usage to the WebUI, even when comparing it's idle usage to their usage under load. Admittedly, that is partially because of my app selection, staying away from eg. gitlab. But still, I don't really get why the WebUI needs so many resources.

**Reply by justinh-rahb** [reply]
My guy, you know you can buy used Xeon CPUs and 8GB sticks of DDR3 ECC memory for less than $10 each right? Not saying everyone should, but if you want to run something that requires it, you sometimes gotta upgrade your hardware. If you prefer newer gen tech, sub-$300 mini-PCs from Amazon come with 16GB RAM and 512GB SSDs nowadays, they make excellent low-power servers on a budget. Raspberry Pi is great for a lot of things, but running WebUI ain't one of them sadly.

**Reply by 9SMTM6** [reply]
I live in the south of germany. Energy is as reliable as it gets, but it's damn expensive.

And I also like to avoid unnessecary waste. Of energy and gear. 

So old gear is out of the question. Compared to a raspberry pi (with 8gb + ssd) most other 'servers' use a few times that energy.

A modernish N100 system might be an option. It's not that much more energy, and it adds processing and compatibility, as well as socketed RAM. But it's still new hardware. And aside of WebUI other services work perfectly well as is.

**Reply by justinh-rahb** [reply]
I've been giving serious thought myself to downsizing my home lab operation significantly. Running decade-old servers is indeed inefficient, can't disagree there. The newer mini-PC based servers are looking pretty good these days, ServeTheHome has an excellent series of articles/videos called Project TinyMiniMicro that goes in depth with using such systems as well as used small-form-factor business PCs as home servers.

**Comment by cheahjs** [comment]
If you strip Open WebUI down, you can get it down to ~200MB of memory use.

The main consumers of memory comes down to all the ML libraries and models that we load. If memory is a concern, then you will want to offload these onto a separate service:

1. Embedding models for RAG. By default OWUI will load an embedding model and the necessary libraries that can consume on the orders of hundreds of MB of RAM. Switch to Ollama or OpenAI-compatible embedding endpoints to offload this task.
2. Whisper model for speech-to-text. Switch to an OpenAI-compatible transcription endpoint to offload this task.

**Reply by FotisK** [reply]
@cheahjs, that's awesome inshifts! Indeed i got the footprint at 225MB, that's much much better! 
Still the VM service for Docker demands 1GB minimum out of the precious GPU/RAM memory.
I think for macOS users a native installation without docker would be much better

**Reply by 9SMTM6** [reply]
I can also confirm that this dropped RAM usage to more like 200MB which, while still more than usual, is perfectly acceptable for me.

**Reply by TedLin1993** [reply]
How can I switch to OpenAI-compatible? I want to deploy in the cloud, but memory over 1GB is very expensive.

**Reply by 9SMTM6** [reply]
Admin settings > audio. I just typed 'invalid' into the token form. But afterwards I tested it and it seems to archive the same reduction with WebAPI. Voice input did not work like that in Firefox, but not like it'll work with an invalid openai token. And not as if I use voice input that often currently.

**Comment by TedLin1993** [comment]
If you use Docker for deployment and want to strip down Open WebUI, you can set the following environment variables:



1. RAG_EMBEDDING_ENGINE: openai

2. AUDIO_STT_ENGINE: openai

3. IMAGE_GENERATION_ENGINE: openai


This will reduce memory usage from approximately 750 MB to around 300 MB. I think this setting should be included in the README.

**Reply by 9SMTM6** [reply]
Great suggestion!

I would default to `RAG_EMBEDDING_ENGINE: ollama`.  `AUDIO_STT_ENGINE` doesnt seem to support the setting webapi, which is available in the UI, so the right approach would be to add it, the lazy approach to go with openai.

`IMAGE_GENERATION_ENGINE` is [disabled by default](https://docs.openwebui.com/getting-started/env-configuration#enable_image_generation) AFAICT, so it doesn't have an effect. If it is enabled manually, then I'd suspect that the person probably will select the choice mostly based on other factors. So maybe mention that, but I'd not put it there by default.

**Reply by 9SMTM6** [reply]
Oh, and it should probably end up on the documentation website. Probably best as a Tutorial.

**Reply by jiangying000** [reply]
good suggestion

**Reply by typed-sigterm** [reply]
After setting that, RAM usage of Open WebUI on my machine increased from 1GB to 1.3GB :( idk why

**Comment by 9SMTM6** [comment]
I've created a documentation PR: https://github.com/open-webui/docs/pull/159

**Comment by kaomoneus** [comment]
I just set RAG_EMBEDDING_ENGINE=ollama as mentioned by @9SMTM6  and @TedLin1993  and it did work to me.

I suspect the reason is that default embedder tries to use RAM instead of VRAM.

**Comment by MarkVader9** [comment]
I mean, no one is saying that you should invest into enterprise solution for exanding your RAM beyond the RAM slots, like  the NVMe SSD
245.76TB Kioxia LC9 and/or the NVMe SSD 122.88TB Solidigm D5-P5336 solutions, but still, if possible, you could expand via NVMe SSD of lower capacity and designate it as a RAM memory for usage (Portainer easiest, or rugged-managed via Docker directly...).
